### PR TITLE
Update no jitpack buld for monorepo

### PR DESCRIPTION
**Pull request by mchowning**
## Description

This PR addresses two issues that were found in the no-jitpack WPAndroid builds with the monorepo:

1. The bundle would not get included in the apk, so the app would crash upon loading the Gutenberg editor; and
2. The build was always rebuilding the bundle even when there were no changes.

In addition, while investigating the always-rebuilding issue, I discovered a problem where the bundle would not get properly generated when a build was run with the clean task. That is also fixed in this PR.

## How has this been tested?

For all of these tests:
1. the WPAndroid app must be built with wp.BUILD_GUTENBERG_FROM_SOURCE either (a) missing or (b) set to false in the `gradle.properties` file. These changes should have no effect when that variable is set to `true`; and
2. The bundler should not be running on the computer. We want to rely solely on the bundle that is included in the apk when testing this PR.

This is just a proposed set of commands that should cover all of the edge cases I have encountered, but it's not a bad idea to try different combinations since that could catch something I missed. All these commands should be run from the monorepo branch for WPAndroid, but with this `fix_no_jitpack_build` branch checked out for gutenberg.

1. Clean the project: `./gradlew clean`
2. Build and install: `./gradlew installWasabiDebug`
3. Run the app and confirm you can open the Gutenberg editor without crashing (establishes that the bundle is being included in the apk)
4. Build the app again with the same command `./gradlew installWasabiDebug`
5. Observe that this build takes under 1 minute whereas the first build took multiple minutes because this build is not regenerating the bundle file (establishes that the build detected there were no changes that would affect the bundle file)
6. Edit any file within the gutenberg-mobile or gutenberg projects that should alter the resulting bundle _other than_ a file within a `node_modules/` folder (the build is just checking the package.json and package-lock.json files for dependency changes).
7. Build the app again with the same command `./gradlew installWasabiDebug`
8. Observe that the app takes multiple minute to build this time because it detected the change and is regenerating the bundle (establishes that the build detected there were no changes that would affect the bundle file).
9. Run clean, build, and install in a single command: `./gradlew clean installWasabiDebug`
10. Run the app and confirm you can open the Gutenberg editor without crashing (establishes does not just check whether the bundle file exists at the beginning of the build, but instead checks it after the "clean" task has run).
11. Try building the `Vanilla` variant. Ideally, do a release build, but if your environment isn't set up for that a debug build is fine: `./gradlew installVanillaRelease`
12. Run the app and confirm you can open the Gutenberg editor without crashing.

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. 
- [X] My code follows the accessibility standards. 
- [X] My code has proper inline documentation. 
- [X] I've included developer documentation if appropriate. 
- [X] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/react-native-bridge/android/build.gradle
Comment (mchowning) [inline packages/react-native-bridge/android/build.gradle:202]: This check has been working for us so far, but I wouldn't be surprised at all if there's a filetype that we've missed. Let me know if you can think of any other "relevant" files we should be checking for changes.

File: packages/react-native-bridge/android/build.gradle
Comment (mchowning) [inline packages/react-native-bridge/android/build.gradle:193]: I'm excluding `react-native-bridge/android/build/intermediates` here because these files were getting generated after a clean, which was throwing off the up-to-date check here since they are js files that are being generated:

```
build/intermediates/library_assets/debug/out/gutenberg-web-single-block/inject-css.js
build/intermediates/library_assets/debug/out/gutenberg-web-single-block/prevent-autosaves.js
build/intermediates/library_assets/debug/out/gutenberg-web-single-block/editor-style-overrides.css
```

My assumption was that we should be fine not checking these since they're generated files, but I'm not really sure what these are, so I wanted to call this out specifically in case it is something we shouldn't be excluding.

Comment (cameronvoell) [reply packages/react-native-bridge/android/build.gradle:193]: Yea these are related to unsupported blocks, and I think this is correct that we don't need to watch for changes on these generated files

File: packages/react-native-bridge/android/build.gradle
Comment (cameronvoell) [reply packages/react-native-bridge/android/build.gradle:202]: Maybe we should include ".scss", or just make it ends with "css" (without the ".")?

Comment (mchowning) [reply packages/react-native-bridge/android/build.gradle]: 👍 Good catch. I made the update. Went with adding `.scss` just for the sake of being super-explicit.

**Comment by cameronvoell** [comment]
This is definitely looking better. One thing I'm seeing, is that after the second build after making a change doesn't seem to skip the **npm_ci / buildJSBundle** task as expected. See steps below:

1. Verify we're in a state with builds working with `./gradlew installWasabiDebug` sub 1 minute because we are not detecting any changes.
2. Make a change to a .js file or something to trigger the extra bundle tasks (longer build)
3. Then the next `./gradlew installWasabiDebug` take a few minutes as expected because it is running the extra tasks. Build works successfully.
4. **Now we run `./gradlew installWasabiDebug` a second time without making any changes, on this second run after a change, I am seeing the `npm_ci` task and `buildJSBundle` tasks run again, making the build take a few minutes. I expected it to already be sub 1 minute without this extra task.**
5. On the third run after the change was made, we are now returning to the sub 1 minute build.

**Comment by mchowning** [comment]
> One thing I'm seeing, is that the second build after making a change doesn't seem to skip the npm_ci / buildJSBundle task as expected.

I'm seeing a couple of things that could be causing this. 

### 1. Package.json formatting changes get undone

One that I just realized is that `npm install` will update the formatting in the `package.json` file, so if the "change" we make is to change the formatting in that file, then the first build sees the change, but subsequently fixes it, and then the second build sees that the formatting has changed, so it redoes everything.

### 2. Package-lock.json is getting modified by the build

A second one is that I am seeing a change being introduced in `gutenberg-mobile/package-lock.json`. The line:

```
"version": "npm:wp-prettier@2.0.5",
```

is being changed to:

```
"version": "npm:prettier@2.0.5",
```

during the build. So the first build sees the original line, then the second build sees that it changed and redoes everything. I'm not seeing the long-second-build issue with these steps:

1. Change something other than `package.json`
2. Insure that `gutenberg-mobile/package-lock.json` has no changes.
3. Run `./gradlew installWasabiDebug` and see the slow build
4. Observe that there are now changes to `gutenberg-mobile/package-lock.json`
5. Revert the `package-lock.json` changes
6. Run `./gradlew installWasabiDebug` and see a fast build.

If undoing the change to `package-lock.json` resolves the issue for you too, I think the issue is with what npm is doing and not with this PR.

**Comment by mchowning** [comment]
> If undoing the change to `package-lock.json` resolves the issue for you too, I think the issue is with what npm is doing and not with this PR. 

Or we could probably just stop checking the package-lock.json files for changes since I believe checking the package.json files should be sufficient.

**Comment by mchowning** [comment]
Pushed 2 updates. 
* 5a60bfe no longer checks package-lock.json files. Instead we're back to just checking the package.json files. That's what we did before, and it seems like that should cover us.
* 5908f93 insures that we are also ignoring changes under the `bundle/android` directory

**Comment by hypest** [comment]
I can confirm that multiple invocations of `./gradlew installWasabiDebug` show that only the first can be long, actually building the JS bundle.

Also, I can confirm that the non-from-source-mode WPAndroid app runs OK and the editor launches OK too without Metro running.

While at it @mchowning , what do you think about revising the ["Building the Gutenberg Mobile JS bundle"](https://github.com/WordPress/gutenberg/blob/47fc9a634ab9d4a1db4653510a485f1d8e443c08/packages/react-native-bridge/android/build.gradle#L45) message in this PR? With all the build optimizations in place, that message is a bit misleading. We could prolly change it to "Executing nested Gutenberg Mobile build tasks..." or something.

**Comment by mchowning** [comment]
> With all the build optimizations in place, that message is a bit misleading. We could prolly change it to "Executing nested Gutenberg Mobile build tasks..." or something.

Certainly up for improving that message, but I feel like if I saw "Executing nested Gutenberg Mobile build tasks..." there's a decent chance I would be misled into thinking that it was a from-source build (`wp.BUILD_GUTENBERG_FROM_SOURCE=true`). I'm trying to see if I can think of of a better message. 

You mentioned that the current message is a bit misleading "with all the build optimizations in place" @hypest . Would you mind explaining a bit how it is misleading? I think that might help me think of some better wording for the message.

**Comment by hypest** [comment]
> You mentioned that the current message is a bit misleading "with all the build optimizations in place" @hypest . Would you mind explaining a bit how it is misleading? I think that might help me think of some better wording for the message.

Oh, so, the current one ("Building the Gutenberg Mobile JS bundle") to me reads like the JS is actually being built, and the JS/RN tools being run. I would expect to see that printed only if the JS bundle is (re)built.

> if I saw "Executing nested Gutenberg Mobile build tasks..." there's a decent chance I would be misled into thinking that it was a from-source build (wp.BUILD_GUTENBERG_FROM_SOURCE=true). I'm trying to see if I can think of of a better message.

Fair.
Actually, since we always do the "build" unless it has been suppressed, perhaps we should remove the message and only print it when suppressed?

Note, this is not a blocker and we can work on it in a follow up PR after the monorepo things get merged.

File: packages/react-native-bridge/android/build.gradle
Comment (hypest) [inline packages/react-native-bridge/android/build.gradle:187]: Is the whitespace change intentional? 🤔
Can we revert it back if not?

Comment (mchowning) [reply packages/react-native-bridge/android/build.gradle]: > Is the whitespace change intentional? 🤔

Not intentional. Weirdly that difference doesn't show up in my IDE or vim when I look at the file. Seems it is a tabs versus spaces issue. Not sure how it got introduced. 😕 

> Can we revert it back if not?

👍  Definitely, done in 90aa339.

**Comment by mchowning** [comment]
Regarding the "Building the Gutenberg Mobile JS bundle" message, I don't feel strongly but unless you have a suggestion you feel good or strongly about @hypest , let's just leave this to be addressed later. I haven't been able to come up with a message I like yet, and there's a decent chance this will be changing a bit anyway in the near future due to jitpack, so that will provide a good opportunity to improve the messaging.

File: packages/react-native-bridge/android/build.gradle
Comment (hypest) [reply packages/react-native-bridge/android/build.gradle]: Sorry Matt, looks like more lines need the same treatment to remove the added whitespace?

Here https://github.com/WordPress/gutenberg/blob/90aa33903d1d398f1f4c15e53f55d9a5f1c8dc14/packages/react-native-bridge/android/build.gradle#L191-L210, and here https://github.com/WordPress/gutenberg/blob/90aa33903d1d398f1f4c15e53f55d9a5f1c8dc14/packages/react-native-bridge/android/build.gradle#L220-L232?

Comment (mchowning) [reply packages/react-native-bridge/android/build.gradle]: Thanks for catching those. No idea why Android Studio has started insisting on tabs instead of spaces despite the fact that I have it set up to use spaces. Anyway, I'm pretty sure I've got them all fixed in 857f407.

**Review by hypest — APPROVED** [review]
Nice fixes @mchowning !

✅ 3. Run the app and confirm you can open the Gutenberg editor without crashing (establishes that the bundle is being included in the apk)

First ^ build took 7m 9s for me.

✅ 5. Observe that this build takes under 1 minute whereas the first build took multiple minutes because this build is not regenerating the bundle file (establishes that the build detected there were no changes that would affect the bundle file)

Second build took 28s for me 🎉

✅ 6. Edit any file within the gutenberg-mobile or gutenberg projects that should alter the resulting bundle other than a file within a node_modules/ folder (the build is just checking the package.json and package-lock.json files for dependency changes).

Added a comment in the root `src/index.js` file. 

✅ Observe that the app takes multiple minute to build this time because it detected the change and is regenerating the bundle (establishes that the build detected there were no changes that would affect the bundle file).

^ took 2m 7s for me.

✅ 10. Run the app and confirm you can open the Gutenberg editor without crashing (establishes does not just check whether the bundle file exists at the beginning of the build, but instead checks it after the "clean" task has run).

^ took 2m 54s for me and app/editor ran OK on device.

✅ 12. Run the app and confirm you can open the Gutenberg editor without crashing.

LGTM! :shipit:

**Comment by hypest** [comment]
Will merge with "preserve commits" now.

**Pull request merged**
